rep-2012.rst

```diff
@@ -0,0 +1,433 @@
+
+   example_interfaces/srv/AddTwoInts_Request[<=1] request
+
+   example_interfaces/srv/AddTwoInts_Response[<=1] response
+
+The definition for ``example_interfaces/srv/AddTwoInts_Request`` is,
+
+.. code-block::
+
+   int64 a
+   int64 b
+
+And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
+
+.. code-block::
+
+   int64 sum
+
+Configuration
+-------------
+
+Configuration of service introspection features will be done through ROS parameters.
+Each node may offer the following four Boolean parameters:
+
+:publish_client_events:
+  If ``true``, then client "request sent" and "response received" events will be published for all clients created by this node.
+  The default value is ``false``.
+:publish_service_events:
+  If ``true``, then service "request received" and "response sent" events will be published for all services created by this node.
+  The default value is ``false``.
+:client_event_payload:
+  If ``true``, then client event messages will have the ``request`` and ``response`` members set.
+  Only applies if ``publish_client_events`` is ``true``.
+  The default value is ``true``.
+:service_event_payload:
+  If ``true``, then service event messages will have the ``request`` and ``response`` members set.
+  Only applies if ``publish_service_events`` is ``true``.
+  The default value is ``true``.
+
+By default, the event publishing feautre is off so users do not pay for a feature they do not plan to use.
+Furthermore, node authors may opt-in by default or disable the service introspection feature altogether as they see fit.
+
+As an example, service introspection can be enabled for all services and clients by providing the following parameters YAML file to ROS nodes [7]_:
+
+.. code-block:: yaml
+
+   /**:
+      ros__parameters:
+         publish_service_events: true
+         publish_client_events: true
+
+Since it is possible to set parameters with a parameter service [8]_, one or more service events *may* be published when setting a service introspection parameter.
+The behavior depends on the order of operations and is defined by the implementation.
+For example, if ``publish_service_events`` was previously set to ``true``, then any service call to change one of the parameters defined above may cause a service request event and/or a response event to be published for the parameter service.
+
+Quality of Service
+------------------
+
+The service event topics proposed in this REP shall use the default quality of service settings [9]_.
+
+Security
+--------
+
+Enabling service introspection creates more attack surface for an existing ROS system by adding 2*N more topics (where N is the number of services with the feature enabled).
+These topics are vulnerable to undesired actors listening in on service communication or even interfering with parts of the system that may be relying on service events.
+
+Luckily, we can leverage the existing security feature for topics and services in ROS 2 (see SROS 2 [10]_).
```

Comment:
Maybe addint a link to the [paper](https://aliasrobotics.com/files/SROS2.pdf) besides the SROS2 repo in here?

Reply:
Done: [027617b](https://github.com/ros-infrastructure/rep/pull/360/commits/027617be2cc3b00798f6fb25b6d5e51f8d2d5c76)